crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,349 @@
+use std::ops::Deref;
+use std::path::PathBuf;
+
+use derive_more::{Deref, DerefMut, Display};
+use flox_types::catalog::cache::{CacheMeta, SubstituterUrl};
+use flox_types::stability::Stability;
+use futures::TryFutureExt;
+use runix::arguments::flake::FlakeArgs;
+use runix::command::Eval;
+use runix::command_line::{NixCommandLine, NixCommandLineRunJsonError};
+use runix::flake_metadata::FlakeMetadata;
+use runix::flake_ref::git::GitRef;
+use runix::flake_ref::indirect::IndirectRef;
+use runix::flake_ref::path::PathRef;
+use runix::flake_ref::{protocol, FlakeRef};
+use runix::installable::{AttrPath, Installable};
+use runix::{RunJson, RunTyped};
+use serde_json::{json, Value};
+use thiserror::Error;
+
+use crate::flox::Flox;
+use crate::providers::git::{GitCommandProvider as Git, GitProvider};
+
+/// Publish state before analyzing
+///
+/// Prevents other actions to commence without analyzing the package first
+pub struct Empty;
+
+/// Publish state after collecting nix metadata
+///
+/// Json value (ideally a [flox_types::catalog::CatalogEntry],
+/// but that's currently broken on account of some flakerefs)
+#[derive(Debug, Deref, DerefMut)]
+pub struct NixAnalysis(Value);
+
+/// State for the publish algorihm
+///
+/// The analysis field tracks the transition from Empty -> NixAnalysis to ensure we don't invoke invalid operations
+pub struct Publish<'flox, State> {
+    /// A shared flox session
+    ///
+    /// Nearly all comands require shared state from the [Flox] object.
+    /// Save a reference to it to simplify the method signatures.
+    flox: &'flox Flox,
+    /// The published _upstream_ source
+    publish_ref: PublishRef,
```

Comment:
Blocker: Need to change this field to something like `publish_flake_ref` (or similar) like we discussed.